### DISCUSSION TITLE
Submitting a Form from a modal shared with multiple child routes and closing the modal on success.

### DISCUSSION THREAD

**Discussion by jbrocher** [remix-run/remix#4694] [Q&A]
Hello there, 

I'm currently building a small App using remix and loving it so far. I am trying to use the remix convention as much as possible, and avoid adding third-party libraries like `Formik` and `react-query`, as I find the remix APIs way more robust. 

However, I've just run into a use case for which I'm having trouble selecting the correct approach. I have a parent route let's call it `/parent/`, below which the user can reach a bunch of nested routes, rendered in `<Outlet>`. One of this nested route is something like `parent/invoices/${id}` this is important for my use case. 

Here are the requirements I want to implement: 

-  I want to render a button in the layout  `/parent/` route, which when clicked, will display a modal with a form for creating an invoice. 
- This modal should appear on top of whatever child of `/parent/` is currently rendering, which isn't necessarily `/parent/invocies`. 
- When the user submits the forms, there are two possibilities: 
    - There are no errors: in that case, the modal should close and the user should be redirected (client-side) to  `parent/invoices/${created_invoice_id}`
    - If there are errors though, the modal should stay open and display them
 - The button and the modal can be used elsewhere in the app. It is a shared component like a newsletter subscription form
 
The main thing here is that the modal cannot be rendered in a nested route like `parent/create-invoice` because this would cause the modal to replace whatever child of `/parent` is currently rendered. (Or any other route, since this component should be shared)
    
 Now I see two possible implementations, but I'm not completely happy with either one
 
 ### Approach 1: Controle modal state with `UseState`
 
I could fallback to client-side logic, with a basic` const [isOpen, setIsOpen] =  useState(false)` to toggle the modal: 
 - When the user clicks on the button set `isOpen` to true
 - Submit the form using useFetcher because the component is shared, and we don't want to cause navigation. 
  - In the action, perform the business logic and either : 
      - Return a response with errors
      - Redirect to `parent/invoices/${created_invoice_id}`
      
 The issue is that, since `parent/invoices/${created_invoice_id}` is a child of `parent`, and the state of the modal is controlled from `parent`, as is, the correct route will be displayed, but the modal will stay open since parent stays mounted and thus the state is kept. 
 
 For now, my solution is to synchronize the state of the modal and the state of the submission with a useEffect: 
 ```
  useEffect(() => {
    if (
      fetcher.type === "done" &&
      !fetcher.data?.error
    ) {
      setIsOpen(false)
    }
  });
```

I'm not happy with this approach: 
 - All the benefits of storing the state of the app in the URL are lost (cannot link to the creation form, close the modal with back navigation and so on)
  - Introducing `useEfffect` feels really forced here and not an idiomatic use of the hook, as we're synchronizing two state variables
  
## Approach 2: Control modal state with a query param 

In this approach, the modal is controlled from a URL, but by a query param, which means it can be added to whichever child of `/parent` is currently rendering. It would work like this: 
 - The button is a GET form that adds a query param to the current route, let's say "?createInvoice=true" 
 - The modal state is controlled by the presence of this parameter, using `useSearchParam`. 
 - When submitting the form, if no errors are present the action will redirect the user to `parent/invoices/${created_invoice_id}` removing `"?createInvoice=true" ` (and any other query param), thus closing the modal
 
 
 This looks way more in line with Remix philosophy however I have a few worries: 
  - Is this approach maintainable in the long term? Repeating this pattern sounds unstable as you in theory open several modals. Maybe this can be solved by using a common `?modal` parameter with the name of the opened modal rather than a parameter for each use case
   - Modifying the query params triggers the loaders. And maintaining `shouldReload` to ignore this para in each loader sounds difficult
   
   
I guess what I am asking is : 
 - What's your opinion on both these approaches? 
 - Is there a better approach I did not think of?
 
EDIT : I've just tested approach n°2 which works fine, but is definitely slower to open the modal than n°1. Probably because a lot of loaders are running. The approach for [Ignoring search params](https://remix.run/docs/en/v1/api/conventions#ignoring-search-params) in the docs seems tedious. If I don't want any loader to load when `?modal` changes, I need to specify `unstable_shouldReload` for each loader, and describe their specific condition for reloading. Ideally, I could tell the loader to ignore `?modal` by default, and make reloading data opt-in (which would make more sense IMO, since you probably now which loader are using the param when you write them, and then you can opt-in)

**Comment by ghost** [comment]
I suggest take advantage of how routes work. 

- Create an Invoice Form and Modal components
- Create a route e.g. /parent/child
- Wrap the content of `child` route inside `<Modal></Modal>`
-- e.g. the `<InvoiceForm />` component you created

To open the modal just visit that route, to close it just navigate away.
E.g. `<Link to="..">Back</Link>` 

In your case perform any validations that you need, when there is an error display a response and, when successful navigate to another route/page. With this approach you can have a `<Link>Form</Link>` anywhere in your application that links user to that modal just style it to look like a button.

**Reply by jbrocher** [reply]
Thanks a lot for your answer :smile: 

However, If I understand you correctly, the modal would live at `parent/child` is that right? If so, when clicking on `<Link>Form</Link>`, the user would navigate away from whichever page is currently rendered. This is fine for most of my requirements, except that I want the modal to be displayed over whatever route is currently rendered. 

This is why in approach n°2 I'm exploring using a search param instead of a dedicated route: So the modal can be displayed in parallel with the currently displayed route.
